### ISSUE TITLE
Optional array assertion does not generate possibly undefined key

### ISSUE THREAD

**Issue by danog** [vimeo/psalm#8647]
https://psalm.dev/r/15fb85c69d

**Comment by orklah** [comment]
Can you post a snippet that does output the possibly undefined key? I'm not sure where you'd expect it

**Comment by danog** [comment]
It's strange, even examples like https://psalm.dev/r/13cb9bff9e don't work properly, a simpler example that works is https://psalm.dev/r/f52ce82e11

**Comment by orklah** [comment]
Yeah, there's a flaw in reconciliation of ifs. You can see this as well here: https://psalm.dev/r/6505682696

I think this may be related

**Comment by weirdan** [comment]
It appears Psalm's inversion of `isset($a['b'])` produces `$a['b'] is null` in the `else` branch. When both branches are reconciled after the `if`, it results in `$a['b'] is null|string`, with the `'b'` key being incorrectly assumed to be defined: https://psalm.dev/r/6721d4dfec
